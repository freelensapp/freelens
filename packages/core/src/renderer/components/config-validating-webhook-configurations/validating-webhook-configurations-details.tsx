/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import type { ValidatingWebhookConfiguration } from "@freelensapp/kube-object";
import { observer } from "mobx-react";
import React from "react";
import { WebhookConfig } from "../config-mutating-webhook-configurations/webhook-config";
import { DrawerItem, DrawerTitle } from "../drawer";
import type { KubeObjectDetailsProps } from "../kube-object-details";

export interface ValidatingWebhookProps extends KubeObjectDetailsProps<ValidatingWebhookConfiguration> {}

@observer
export class ValidatingWebhookDetails extends React.Component<ValidatingWebhookProps> {
  render() {
    const { object: webhookConfig } = this.props;

    return (
      <div className="ValidatingWebhookDetails">
        <DrawerItem name="API version">{webhookConfig.apiVersion}</DrawerItem>
        <DrawerTitle>Webhooks</DrawerTitle>
        {webhookConfig.getWebhooks()?.length == 0 && <div style={{ opacity: 0.6 }}>No webhooks set</div>}
        {webhookConfig.getWebhooks()?.map((webhook) => (
          <WebhookConfig webhook={webhook} key={webhook.name} />
        ))}
      </div>
    );
  }
}
