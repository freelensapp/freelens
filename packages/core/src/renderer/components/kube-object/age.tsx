/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import React from "react";
import { ReactiveDuration } from "../duration/reactive-duration";

export interface KubeObjectAgeProps {
  object: {
    metadata: {
      creationTimestamp?: string;
    };
  };

  /**
   * Whether the display string should prefer length over precision
   * @default true
   */
  compact?: boolean;
}

export const KubeObjectAge = ({ object, compact = true }: KubeObjectAgeProps) => (
  <ReactiveDuration timestamp={object.metadata.creationTimestamp} compact={compact} />
);
