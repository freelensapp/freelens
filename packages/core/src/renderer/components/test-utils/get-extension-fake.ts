/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import { LensMainExtension } from "../../../extensions/lens-main-extension";
import { LensRendererExtension } from "../../../extensions/lens-renderer-extension";

export class TestExtensionMain extends LensMainExtension {}
export class TestExtensionRenderer extends LensRendererExtension {}

export interface FakeExtensionOptions {
  id: string;
  name: string;
  rendererOptions?: Partial<LensRendererExtension>;
  mainOptions?: Partial<LensMainExtension>;
}

export const getExtensionFakeForMain = ({ id, name, mainOptions = {} }: FakeExtensionOptions) =>
  Object.assign(
    new TestExtensionMain({
      id,
      absolutePath: "irrelevant",
      isBundled: false,
      isCompatible: false,
      isEnabled: false,
      manifest: {
        name,
        version: "1.0.0",
        engines: {
          freelens: "^0.1.0",
        },
      },
      manifestPath: "irrelevant",
    }),
    mainOptions,
  );

export const getExtensionFakeForRenderer = ({ id, name, rendererOptions = {} }: FakeExtensionOptions) =>
  Object.assign(
    new TestExtensionRenderer({
      id,
      absolutePath: "irrelevant",
      isBundled: false,
      isCompatible: false,
      isEnabled: false,
      manifest: {
        name,
        version: "1.0.0",
        engines: {
          freelens: "^0.1.0",
        },
      },
      manifestPath: "irrelevant",
    }),
    rendererOptions,
  );
