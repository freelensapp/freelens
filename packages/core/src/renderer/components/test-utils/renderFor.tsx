/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import type { DiContainer } from "@ogre-tools/injectable";
import { DiContextProvider } from "@ogre-tools/injectable-react";
import type { RenderResult } from "@testing-library/react";
import { render as testingLibraryRender } from "@testing-library/react";
import React from "react";

export type DiRender = (ui: React.ReactElement) => RenderResult;

type DiRenderFor = (di: DiContainer) => DiRender;

export const renderFor: DiRenderFor = (di) => (ui) => {
  const result = testingLibraryRender(<DiContextProvider value={{ di }}>{ui}</DiContextProvider>);

  return {
    ...result,

    rerender: (ui: React.ReactElement) => result.rerender(<DiContextProvider value={{ di }}>{ui}</DiContextProvider>),
  };
};
