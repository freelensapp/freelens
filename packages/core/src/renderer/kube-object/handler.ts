/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import type { BaseIconProps } from "@freelensapp/icon";
import type { KubeObject } from "@freelensapp/kube-object";
import { getInjectionToken } from "@ogre-tools/injectable";
import type { IObservableArray } from "mobx";
import type { RequireAtLeastOne } from "type-fest";

export interface KubeObjectContextMenuItem {
  id?: string;
  /**
   * If the type is `string` then it is shorthand for {@link BaseIconProps.material}
   *
   * This is required because this item can be either rendered as a context menu or as a toolbar in
   * the kube object details page.
   */
  icon: string | BaseIconProps;

  /**
   * The title text for the menu item or the hover text for the icon.
   */
  title: string;

  /**
   * The action when clicked
   */
  onClick: (obj: KubeObject) => void;
}

export interface KubeObjectOnContextMenuOpenContext {
  readonly menuItems: IObservableArray<KubeObjectContextMenuItem>;
  navigate: (location: string) => void;
}

export type KubeObjectOnContextMenuOpen = (ctx: KubeObjectOnContextMenuOpenContext) => void;

export interface KubeObjectHandlers {
  onContextMenuOpen: KubeObjectOnContextMenuOpen;
}

export type KubeObjectHandlerRegistration = {
  apiVersions: string[];
  kind: string;
} & RequireAtLeastOne<KubeObjectHandlers>;

export const staticKubeObjectHandlerInjectionToken = getInjectionToken<KubeObjectHandlerRegistration>({
  id: "static-kube-object-handler-token",
});
