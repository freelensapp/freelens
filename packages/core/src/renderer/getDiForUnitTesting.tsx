/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import { animateFeature, requestAnimationFrameInjectable } from "@freelensapp/animate";
import { clusterSidebarFeature } from "@freelensapp/cluster-sidebar";
import { registerFeature } from "@freelensapp/feature-core";
import { kubeApiSpecificsFeature } from "@freelensapp/kube-api-specifics";
import { setLegacyGlobalDiForExtensionApi } from "@freelensapp/legacy-global-di";
import { loggerFeature } from "@freelensapp/logger";
import { messagingFeature, testUtils as messagingTestUtils } from "@freelensapp/messaging";
import { notificationsFeature } from "@freelensapp/notifications";
import { randomFeature } from "@freelensapp/random";
import { routingFeature } from "@freelensapp/routing";
import type { GlobalOverride } from "@freelensapp/test-utils";
import { createContainer, isInjectable } from "@ogre-tools/injectable";
import { registerMobX } from "@ogre-tools/injectable-extension-for-mobx";
import { registerInjectableReact } from "@ogre-tools/injectable-react";
import { chunk, noop } from "lodash/fp";
import { runInAction } from "mobx";
import { getOverrideFsWithFakes } from "../test-utils/override-fs-with-fakes";
import hostedClusterIdInjectable from "./cluster-frame-context/hosted-cluster-id.injectable";
import terminalSpawningPoolInjectable from "./components/dock/terminal/terminal-spawning-pool.injectable";
import startTopbarStateSyncInjectable from "./components/layout/top-bar/start-state-sync.injectable";
import legacyOnChannelListenInjectable from "./ipc/legacy-channel-listen.injectable";
import watchHistoryStateInjectable from "./remote-helpers/watch-history-state.injectable";

export const getDiForUnitTesting = () => {
  const environment = "renderer";
  const di = createContainer(environment, {
    detectCycles: false,
  });

  registerMobX(di);
  registerInjectableReact(di);
  setLegacyGlobalDiForExtensionApi(di, environment);

  runInAction(() => {
    registerFeature(
      di,
      messagingFeature,
      messagingTestUtils.messagingFeatureForUnitTesting,
      routingFeature,
      loggerFeature,
      animateFeature,
      clusterSidebarFeature,
      randomFeature,
      kubeApiSpecificsFeature,
      notificationsFeature,
    );
  });

  di.preventSideEffects();

  runInAction(() => {
    const injectables = global.injectablePaths.renderer.paths
      .map((path) => require(path))
      .flatMap(Object.values)
      .filter(isInjectable);

    for (const block of chunk(100)(injectables)) {
      di.register(...block);
    }
  });

  for (const globalOverridePath of global.injectablePaths.renderer.globalOverridePaths) {
    const globalOverride = require(globalOverridePath).default as GlobalOverride<unknown, unknown, unknown>;

    di.override(globalOverride.injectable, globalOverride.overridingInstantiate);
  }

  [startTopbarStateSyncInjectable].forEach((injectable) => {
    di.override(injectable, () => ({
      id: injectable.id,
      run: () => {},
    }));
  });

  di.override(terminalSpawningPoolInjectable, () => document.createElement("div"));
  di.override(hostedClusterIdInjectable, () => undefined);

  di.override(legacyOnChannelListenInjectable, () => () => noop);

  di.override(requestAnimationFrameInjectable, () => (callback) => callback());
  di.override(watchHistoryStateInjectable, () => () => () => {});

  getOverrideFsWithFakes()(di);

  return di;
};
