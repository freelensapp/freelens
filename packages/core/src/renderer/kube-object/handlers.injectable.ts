/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import { getOrInsert, getOrInsertMap, readonly } from "@freelensapp/utilities";
import { getInjectable } from "@ogre-tools/injectable";
import { computed } from "mobx";
import rendererExtensionsInjectable from "../../extensions/renderer-extensions.injectable";
import type { KubeObjectHandlerRegistration, KubeObjectHandlers } from "./handler";
import { staticKubeObjectHandlerInjectionToken } from "./handler";

const kubeObjectHandlersInjectable = getInjectable({
  id: "kube-object-handlers",
  instantiate: (di) => {
    const extensions = di.inject(rendererExtensionsInjectable);
    const staticKubeObjectContextMenuHandlers = di.injectMany(staticKubeObjectHandlerInjectionToken);

    return computed(() => {
      const res = new Map<string, Map<string, Partial<KubeObjectHandlers>[]>>();
      const addAllHandlers = (registrations: KubeObjectHandlerRegistration[]) => {
        for (const { apiVersions, kind, ...handlers } of registrations) {
          for (const apiVersion of apiVersions) {
            const byApiVersion = getOrInsertMap(res, apiVersion);
            const byKind = getOrInsert(byApiVersion, kind, []);

            byKind.push(handlers);
          }
        }
      };

      extensions
        .get()
        .map((ext) => ext.kubeObjectHandlers)
        .forEach(addAllHandlers);

      addAllHandlers(staticKubeObjectContextMenuHandlers);

      return readonly(res);
    });
  },
});

export default kubeObjectHandlersInjectable;
