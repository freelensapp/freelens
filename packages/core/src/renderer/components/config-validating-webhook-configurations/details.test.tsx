/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import type { ValidatingWebhookConfigurationData } from "@freelensapp/kube-object";
import { ValidatingWebhookConfiguration } from "@freelensapp/kube-object";
import type { RenderResult } from "@testing-library/react";
import React from "react";
import { getDiForUnitTesting } from "../../getDiForUnitTesting";
import type { DiRender } from "../test-utils/renderFor";
import { renderFor } from "../test-utils/renderFor";
import { ValidatingWebhookDetails } from "./validating-webhook-configurations-details";

const validatingWebhookConfig: ValidatingWebhookConfigurationData = {
  apiVersion: "admissionregistration.k8s.io/v1",
  kind: "ValidatingWebhookConfiguration",
  metadata: {
    name: "pod-policy.example.com",
    resourceVersion: "1",
    uid: "pod-policy.example.com",
    namespace: "default",
    selfLink: "/apis/admissionregistration.k8s.io/v1/pod-policy.example.com",
  },
  webhooks: [
    {
      name: "pod-policy.example.com",
      rules: [
        {
          apiGroups: ["", "apps", "extensions"],
          apiVersions: ["v1", "v1beta1"],
          operations: ["CREATE", "UPDATE"],
          resources: ["pods", "deployments"],
        },
      ],
      failurePolicy: "Fail",
      admissionReviewVersions: ["v1", "v1beta1"],
      sideEffects: "None",
      timeoutSeconds: 5,
      clientConfig: {
        service: {
          namespace: "service-namespace",
          name: "service-name",
        },
        caBundle: "Cg==",
      },
    },
  ],
};

describe("ValidatingWebhookConfigsDetails", () => {
  let result: RenderResult;
  let render: DiRender;

  beforeEach(() => {
    const di = getDiForUnitTesting();

    render = renderFor(di);
  });

  it("renders", () => {
    const webhookConfig = new ValidatingWebhookConfiguration(validatingWebhookConfig);

    result = render(<ValidatingWebhookDetails object={webhookConfig} />);

    expect(result.baseElement).toMatchSnapshot();
  });

  it("renders with no webhooks", () => {
    const webhookConfig = new ValidatingWebhookConfiguration({
      ...validatingWebhookConfig,
      webhooks: [],
    });

    result = render(<ValidatingWebhookDetails object={webhookConfig} />);

    expect(result.baseElement).toMatchSnapshot();
  });
});
