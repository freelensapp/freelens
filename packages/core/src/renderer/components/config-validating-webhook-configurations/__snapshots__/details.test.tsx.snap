// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValidatingWebhookConfigsDetails renders 1`] = `
<body>
  <div>
    <div
      class="ValidatingWebhookDetails"
    >
      <div
        class="DrawerItem"
      >
        <span
          class="name"
        >
          API version
        </span>
        <span
          class="value"
        >
          admissionregistration.k8s.io/v1
        </span>
      </div>
      <div
        class="DrawerTitle title"
      >
        Webhooks
      </div>
      <div
        class="DrawerItem firstItem"
      >
        <span
          class="name"
        >
          Name
        </span>
        <span
          class="value"
        >
          <strong>
            pod-policy.example.com
          </strong>
        </span>
      </div>
      <div
        class="DrawerItem"
      >
        <span
          class="name"
        >
          Client Config
        </span>
        <span
          class="value"
        >
          <div>
            <div>
              Name: 
              service-name
            </div>
            <div>
              Namespace: 
              service-namespace
            </div>
          </div>
        </span>
      </div>
      <div
        class="DrawerItem"
      >
        <span
          class="name"
        >
          Match Policy
        </span>
        <span
          class="value"
        />
      </div>
      <div
        class="DrawerItem"
      >
        <span
          class="name"
        >
          Failure Policy
        </span>
        <span
          class="value"
        >
          Fail
        </span>
      </div>
      <div
        class="DrawerItem"
      >
        <span
          class="name"
        >
          Admission Review Versions
        </span>
        <span
          class="value"
        >
          v1, v1beta1
        </span>
      </div>
      <div
        class="DrawerItem"
      >
        <span
          class="name"
        >
          Side Effects
        </span>
        <span
          class="value"
        >
          None
        </span>
      </div>
      <div
        class="DrawerItem"
      >
        <span
          class="name"
        >
          Timeout Seconds
        </span>
        <span
          class="value"
        >
          5
        </span>
      </div>
      <div
        class="DrawerItem"
      >
        <span
          class="name"
        >
          Namespace Selector
        </span>
        <span
          class="value"
        />
      </div>
      <div
        class="DrawerItem"
      >
        <span
          class="name"
        >
          Object Selector
        </span>
        <span
          class="value"
        />
      </div>
      <div
        class="DrawerItem lastItem"
      >
        <span
          class="name"
        >
          Rules
        </span>
        <span
          class="value"
        >
          <div>
            <div>
              API Groups: 
              , apps, extensions
            </div>
            <div>
              API Versions: 
              v1, v1beta1
            </div>
            <div>
              Operations: 
              CREATE, UPDATE
            </div>
            <div>
              Resources: 
              pods, deployments
            </div>
          </div>
        </span>
      </div>
    </div>
  </div>
</body>
`;

exports[`ValidatingWebhookConfigsDetails renders with no webhooks 1`] = `
<body>
  <div>
    <div
      class="ValidatingWebhookDetails"
    >
      <div
        class="DrawerItem"
      >
        <span
          class="name"
        >
          API version
        </span>
        <span
          class="value"
        >
          admissionregistration.k8s.io/v1
        </span>
      </div>
      <div
        class="DrawerTitle title"
      >
        Webhooks
      </div>
      <div
        style="opacity: 0.6;"
      >
        No webhooks set
      </div>
    </div>
  </div>
</body>
`;
